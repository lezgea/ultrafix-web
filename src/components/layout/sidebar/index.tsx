import React from 'react';

interface ISidebarProps {
    navLinks: React.ReactNode[];
    visible: boolean;
    setSidebarOpen: (val: boolean) => void;
}

export const Sidebar: React.FC<ISidebarProps> = ({ navLinks, visible, setSidebarOpen }) => {
    const sidebarRef = React.useRef<HTMLDivElement>(null);

    React.useEffect(() => {
        const handleClickOutside = (event: MouseEvent) => {
            if (sidebarRef.current && !sidebarRef.current.contains(event.target as Node)) {
                setSidebarOpen(false);
            }
        };

        document.addEventListener('mousedown', handleClickOutside);
        return () => {
            document.removeEventListener('mousedown', handleClickOutside);
        };
    }, [setSidebarOpen]);

    return (
        visible && (
            <div
                data-testid="sidebar"
                className="fixed inset-0 z-20 bg-gray-800 bg-opacity-75 top-[65px] transition-transform transform translate-x-0 opacity-1 lg:hidden"
            >
                <div
                    className="relative w-64 h-[100%] bg-white shadow-xl py-8"
                    ref={sidebarRef}
                    onClick={(e) => e.stopPropagation()} // Prevent event propagation
                >
                    <nav>
                        <ul className="space-y-6 px-8">
                            {navLinks}
                        </ul>
                    </nav>
                </div>
            </div>
        )
    );
};
