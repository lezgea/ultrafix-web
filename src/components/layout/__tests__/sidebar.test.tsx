import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Sidebar } from '../sidebar';

describe('Sidebar Component', () => {
    const mockNavLinks = [
        <li key="about">About Us</li>,
        <li key="races">Races</li>,
        <li key="faq">FAQ</li>,
        <li key="contact">Contact</li>,
    ];

    const mockSetSidebarOpen = jest.fn();

    beforeEach(() => {
        mockSetSidebarOpen.mockClear(); // Clear the mock function before each test
    });

    it('should render the sidebar when visible is true', () => {
        const { asFragment } = render(
            <Sidebar navLinks={mockNavLinks} visible={true} setSidebarOpen={mockSetSidebarOpen} />
        );

        expect(screen.getByTestId('sidebar')).toBeInTheDocument();
        expect(screen.getByText('About Us')).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should close the sidebar when clicking outside of it', () => {
        const { asFragment } = render(
            <Sidebar navLinks={mockNavLinks} visible={true} setSidebarOpen={mockSetSidebarOpen} />
        );

        fireEvent.mouseDown(document);

        // Assert that setSidebarOpen was called with false
        expect(mockSetSidebarOpen).toHaveBeenCalledWith(false);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correctly when sidebar is not visible', () => {
        const { asFragment } = render(
            <Sidebar navLinks={mockNavLinks} visible={false} setSidebarOpen={mockSetSidebarOpen} />
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
