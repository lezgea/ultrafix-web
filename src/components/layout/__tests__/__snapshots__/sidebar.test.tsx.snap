// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Component should close the sidebar when clicking outside of it 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 z-20 bg-gray-800 bg-opacity-75 top-[65px] transition-transform transform translate-x-0 opacity-1 lg:hidden"
    data-testid="sidebar"
  >
    <div
      class="relative w-64 h-[100%] bg-white shadow-xl py-8"
    >
      <nav>
        <ul
          class="space-y-6 px-8"
        >
          <li>
            About Us
          </li>
          <li>
            Races
          </li>
          <li>
            FAQ
          </li>
          <li>
            Contact
          </li>
        </ul>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Sidebar Component should render correctly when sidebar is not visible 1`] = `<DocumentFragment />`;

exports[`Sidebar Component should render the sidebar when visible is true 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 z-20 bg-gray-800 bg-opacity-75 top-[65px] transition-transform transform translate-x-0 opacity-1 lg:hidden"
    data-testid="sidebar"
  >
    <div
      class="relative w-64 h-[100%] bg-white shadow-xl py-8"
    >
      <nav>
        <ul
          class="space-y-6 px-8"
        >
          <li>
            About Us
          </li>
          <li>
            Races
          </li>
          <li>
            FAQ
          </li>
          <li>
            Contact
          </li>
        </ul>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;
